File: src/test/java/dev/kikugie/techutils/test/GiveFullInvTest.java
package dev.kikugie.techutils.test;

import dev.kikugie.techutils.feature.GiveFullIInv;
import net.minecraft.Bootstrap;
import net.minecraft.SharedConstants;
import net.minecraft.block.Block;
import net.minecraft.block.Blocks;
import net.minecraft.block.ShulkerBoxBlock;
import net.minecraft.block.entity.ChestBlockEntity;
import net.minecraft.block.entity.ShulkerBoxBlockEntity;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.BundleContentsComponent;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.util.DyeColor;
import net.minecraft.util.math.BlockPos;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import dev.kikugie.techutils.config.MiscConfigs;

public class GiveFullInvTest {
    @BeforeAll
    public static void setup() {
        SharedConstants.createGameVersion();
        Bootstrap.initialize();
    }

    /**
     * <pre>
     *     Main hand: 1 diamond
     *     Off hand: none
     *     Expected result: 1 shulker box (1728 diamonds)</pre>
     */
    @Test
    public void itemInBox() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullBox = GiveFullIInv.get(testItem, ItemStack.EMPTY);
        Assertions.assertTrue(fullBox.isPresent(), "Failed to insert item in a box");
        boxOf(testItem.copyWithCount(64), fullBox.get(), null);
    }

    /**
     * <pre>
     *     Main hand: 1 diamond
     *     Off hand: 1 white shulker box
     *     Expected result: 1 white shulker box (1728 diamonds)</pre>
     */
    @Test
    public void itemInColoredBox() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullBox = GiveFullIInv.get(testItem, getEmptyBox(DyeColor.WHITE));
        Assertions.assertTrue(fullBox.isPresent(), "Failed to insert item in a box");
        boxOf(testItem.copyWithCount(64), fullBox.get(), DyeColor.WHITE);
    }

    /**
     * <pre>
     *     Main hand: 1 diamond
     *     Off hand: 1 chest
     *     Expected result: 1 chest (1728 diamonds)</pre>
     */
    @Test
    public void itemInChest() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullChest = GiveFullIInv.get(testItem, getEmptyChest());
        Assertions.assertTrue(fullChest.isPresent(), "Failed to insert item in a chest");
        chestOf(testItem.copyWithCount(64), fullChest.get());
    }

    /**
     * <pre>
     *     Main hand: 1 diamond
     *     Off hand: 1 bundle
     *     Expected result: 1 bundle (64 * {@link MiscConfigs#BUNDLE_FILL} diamonds)</pre>
     */
    @Test
    public void itemInBundle() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullBundle = GiveFullIInv.get(testItem, getEmptyBundle());
        Assertions.assertTrue(fullBundle.isPresent(), "Failed to insert item in a bundle");
        bundleOf(testItem.copyWithCount(64), fullBundle.get());
    }

    /**
     * <pre>
     *     Main hand: 1 shulker box (1728 diamonds)
     *     Off hand: none
     *     Expected result: 1 chest (27 shulker boxes)</pre>
     */
    @Test
    public void boxInChest() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullBox = GiveFullIInv.get(testItem, ItemStack.EMPTY);
        Assertions.assertTrue(fullBox.isPresent(), "Failed to insert item in a box");

        Optional<ItemStack> fullChest = GiveFullIInv.get(fullBox.get(), getEmptyChest());
        Assertions.assertTrue(fullChest.isPresent(), "Failed to insert boxes in a chest");
        chestOf(fullBox.get(), fullChest.get());
    }

    /**
     * <pre>
     *     Main hand: 1 shulker box (1728 diamonds)
     *     Off hand: 1 bundle
     *     Expected result: 1 bundle (64 * {@link MiscConfigs#BUNDLE_FILL} shulker boxes)</pre>
     */
    @Test
    public void boxInBundle() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullBox = GiveFullIInv.get(testItem, ItemStack.EMPTY);
        Assertions.assertTrue(fullBox.isPresent(), "Failed to insert item in a box");

        Optional<ItemStack> fullBundle = GiveFullIInv.get(fullBox.get(), getEmptyBundle());
        Assertions.assertTrue(fullBundle.isPresent(), "Failed to insert boxes in a chest");
        bundleOf(fullBox.get(), fullBundle.get());
    }

    /**
     * <pre>
     *     Main hand: 1 shulker box (1728 diamonds)
     *     Off hand: 1 shulker box
     *     Expected result: none</pre>
     */
    @Test
    public void boxInBox() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullBox = GiveFullIInv.get(testItem, ItemStack.EMPTY);
        Assertions.assertTrue(fullBox.isPresent(), "Failed to insert item in a box");

        Optional<ItemStack> fullBox2 = GiveFullIInv.get(fullBox.get(), getEmptyBox(null));
        Assertions.assertFalse(fullBox2.isPresent(), "Should return no item");
    }

    /**
     * <pre>
     *     Main hand: 1 white shulker box
     *     Off hand: 1 chest
     *     Expected result: 1 chest (1728 shulker boxes)</pre>
     */
    @Test
    public void emptyBoxInChest() {
        ItemStack testItem = getEmptyBox(DyeColor.WHITE);
        Optional<ItemStack> fullChest = GiveFullIInv.get(testItem, getEmptyChest());
        Assertions.assertTrue(fullChest.isPresent(), "Failed to insert boxes in a chest");
        chestOf(testItem.copyWithCount(64), fullChest.get());
    }

    /**
     * <pre>
     *     Main hand: 1 white shulker box
     *     Off hand: none
     *     Expected result: 1 chest (1728 shulker boxes)</pre>
     */
    @Test
    public void emptyBox() {
        ItemStack testItem = getEmptyBox(DyeColor.WHITE);
        Optional<ItemStack> fullChest = GiveFullIInv.get(testItem, ItemStack.EMPTY);
        Assertions.assertTrue(fullChest.isPresent(), "Failed to insert boxes in a chest");
        chestOf(testItem.copyWithCount(64), fullChest.get());
    }

    /**
     * <pre>
     *     Main hand: 1 white shulker box
     *     Off hand: 1 bundle
     *     Expected result: 1 bundle (64 * {@link MiscConfigs#BUNDLE_FILL} shulker boxes)</pre>
     */
    @Test
    public void emptyBoxInBundle() {
        ItemStack testItem = getEmptyBox(DyeColor.WHITE);
        Optional<ItemStack> fullBundle = GiveFullIInv.get(testItem, getEmptyBundle());
        Assertions.assertTrue(fullBundle.isPresent(), "Failed to insert boxes in a bundle");
        bundleOf(testItem.copyWithCount(64), fullBundle.get());
    }

    /**
     * <pre>
     *     Main hand: 1 white shulker box
     *     Off hand: 1 shulker box
     *     Expected result: none</pre>
     */
    @Test
    public void emptyBoxInBox() {
        ItemStack testItem = getEmptyBox(DyeColor.WHITE);
        Optional<ItemStack> fullBox = GiveFullIInv.get(testItem, getEmptyBox(null));
        Assertions.assertFalse(fullBox.isPresent(), "Should return no item");
    }

    /**
     * <pre>
     *     Main hand: none
     *     Off hand: none
     *     Expected result: none</pre>
     */
    @Test
    public void noItem() {
        Optional<ItemStack> fullBox = GiveFullIInv.get(ItemStack.EMPTY, ItemStack.EMPTY);
        Assertions.assertFalse(fullBox.isPresent(), "Should return no item");
    }

    /**
     * <pre>
     *     Main hand: none
     *     Off hand: 1 shulker box
     *     Expected result: none</pre>
     */
    @Test
    public void noItemWithOffhand() {
        Optional<ItemStack> fullBox = GiveFullIInv.get(ItemStack.EMPTY, getEmptyBox(null));
        Assertions.assertFalse(fullBox.isPresent(), "Should return no item");
    }

    /**
     * <pre>
     *     Main hand: 1 chest (1728 diamonds)
     *     Off hand: 1 shulker box
     *     Expected result: none</pre>
     */
    @Test
    public void chestInBoxInOffhand() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullChest = GiveFullIInv.get(testItem, getEmptyChest());
        Assertions.assertTrue(fullChest.isPresent(), "Failed to insert item in a box");

        Optional<ItemStack> fullBox = GiveFullIInv.get(fullChest.get(), getEmptyBox(null));
        Assertions.assertFalse(fullBox.isPresent(), "Should return no item");
    }

    /**
     * <pre>
     *     Main hand: 1 chest (1728 diamonds)
     *     Off hand: none
     *     Expected result: none</pre>
     */
    @Test
    public void chestInBox() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullChest = GiveFullIInv.get(testItem, getEmptyChest());
        Assertions.assertTrue(fullChest.isPresent(), "Failed to insert item in a box");

        Optional<ItemStack> fullBox = GiveFullIInv.get(fullChest.get(), ItemStack.EMPTY);
        Assertions.assertFalse(fullBox.isPresent(), "Should return no item");
    }

    /**
     * <pre>
     *     Main hand: 1 chest (1728 diamonds)
     *     Off hand: 1 bundle
     *     Expected result: none</pre>
     */
    @Test
    public void chestInBundle() {
        ItemStack testItem = getTestItem();
        Optional<ItemStack> fullChest = GiveFullIInv.get(testItem, getEmptyChest());
        Assertions.assertTrue(fullChest.isPresent(), "Failed to insert item in a box");

        Optional<ItemStack> fullBundle = GiveFullIInv.get(fullChest.get(), getEmptyBundle());
        Assertions.assertFalse(fullBundle.isPresent(), "Should return no item");
    }

    private void boxOf(ItemStack stack, ItemStack box, @Nullable DyeColor color) {
        Assertions.assertEquals(color, getBoxColor(box.getItem()), "Shulker box color '%s' doesn't match expected '%s'".formatted(getBoxColor(box.getItem()), color));
		Assertions.assertNotNull(box.get(DataComponentTypes.CONTAINER), "Shulker box has no container component");

		ShulkerBoxBlockEntity shulker = new ShulkerBoxBlockEntity(color, BlockPos.ORIGIN, ShulkerBoxBlock.get(color).getDefaultState());
        shulker.readComponents(stack);
        for (int i = 0; i < shulker.size(); i++) {
            Assertions.assertTrue(ItemStack.areEqual(stack, shulker.getStack(i)), "Shulker box item '%s' doesn't match expected '%s'".formatted(shulker.getStack(i), stack));
        }
    }

    private void chestOf(ItemStack stack, ItemStack chest) {
		Assertions.assertNotNull(chest.get(DataComponentTypes.CONTAINER), "Chest has no container component");

        ChestBlockEntity container = new ChestBlockEntity(BlockPos.ORIGIN, Blocks.CHEST.getDefaultState());
        container.readComponents(chest);
        for (int i = 0; i < container.size(); i++) {
            Assertions.assertTrue(ItemStack.areEqual(stack, container.getStack(i)), "Chest item '%s' doesn't match expected '%s'".formatted(container.getStack(i), stack));
        }
    }

    private void bundleOf(ItemStack stack, ItemStack bundle) {
        BundleContentsComponent contents = bundle.get(DataComponentTypes.BUNDLE_CONTENTS);
        Assertions.assertNotNull(contents, "Bundle has no contents component");

        Assertions.assertFalse(contents.isEmpty(), "Bundle is empty");

        for (ItemStack item : contents.iterate()) {
            Assertions.assertTrue(ItemStack.areEqual(stack, item), "Bundle item '%s' doesn't match expected '%s'".formatted(item, stack));
        }
    }

    private ItemStack getTestItem() {
        return Items.DIAMOND.getDefaultStack();
    }

    private ItemStack getEmptyBox(@Nullable DyeColor color) {
        return ShulkerBoxBlock.getItemStack(color);
    }

    private ItemStack getEmptyChest() {
        return Items.CHEST.getDefaultStack();
    }

    private ItemStack getEmptyDropper() {
        return Items.DROPPER.getDefaultStack();
    }

    private ItemStack getEmptyBundle() {
        return Items.BUNDLE.getDefaultStack();
    }

    @Nullable
    private static DyeColor getBoxColor(Item item) {
        return getBoxColor(Block.getBlockFromItem(item));
    }

    @Nullable
    private static DyeColor getBoxColor(Block block) {
        return block instanceof ShulkerBoxBlock box ? box.getColor() : null;
    }
}

File: src/test/kotlin/dev/kikugie/techutils/test/GiveFullInvTest.kt
package dev.kikugie.techutils.test

import dev.kikugie.techutils.client.feature.misc.impl.GiveFullIInv.getStack
import net.minecraft.Bootstrap
import net.minecraft.SharedConstants
import net.minecraft.block.Blocks
import net.minecraft.block.ShulkerBoxBlock
import net.minecraft.block.entity.ChestBlockEntity
import net.minecraft.block.entity.ShulkerBoxBlockEntity
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtElement
import net.minecraft.util.DyeColor
import net.minecraft.util.math.BlockPos
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

class GiveFullInvTest {
    /**
     * <pre>
     * Main hand: 1 diamond
     * Off hand: none
     * Expected result: 1 shulker box (1728 diamonds)</pre>
     */
    @Test
    fun itemInBox() {
        val testItem = testItem
        val fullBox: ItemStack? = getStack(testItem, ItemStack.EMPTY)
        Assertions.assertTrue(fullBox != null, "Failed to insert item in a box")
        boxOf(testItem.copyWithCount(64), fullBox!!, null)
    }

    /**
     * <pre>
     * Main hand: 1 diamond
     * Off hand: 1 white shulker box
     * Expected result: 1 white shulker box (1728 diamonds)</pre>
     */
    @Test
    fun itemInColoredBox() {
        val testItem = testItem
        val fullBox: ItemStack? = getStack(testItem, getEmptyBox(DyeColor.WHITE))
        Assertions.assertTrue(fullBox != null, "Failed to insert item in a box")
        boxOf(testItem.copyWithCount(64), fullBox!!, DyeColor.WHITE)
    }

    /**
     * <pre>
     * Main hand: 1 diamond
     * Off hand: 1 chest
     * Expected result: 1 chest (1728 diamonds)</pre>
     */
    @Test
    fun itemInChest() {
        val testItem = testItem
        val fullChest: ItemStack? = getStack(testItem, emptyChest)
        Assertions.assertTrue(fullChest != null, "Failed to insert item in a chest")
        chestOf(testItem.copyWithCount(64), fullChest!!)
    }

    /**
     * <pre>
     * Main hand: 1 diamond
     * Off hand: 1 bundle
     * Expected result: 1 bundle (64 *  diamonds)</pre>
     */
    @Test
    fun itemInBundle() {
        val testItem = testItem
        val fullBundle: ItemStack? = getStack(testItem, emptyBundle)
        Assertions.assertTrue(fullBundle != null, "Failed to insert item in a bundle")
        bundleOf(testItem.copyWithCount(64), fullBundle!!)
    }

    /**
     * <pre>
     * Main hand: 1 shulker box (1728 diamonds)
     * Off hand: none
     * Expected result: 1 chest (27 shulker boxes)</pre>
     */
    @Test
    fun boxInChest() {
        val testItem = testItem
        val fullBox: ItemStack? = getStack(testItem, ItemStack.EMPTY)
        Assertions.assertTrue(fullBox != null, "Failed to insert item in a box")
        val fullChest: ItemStack? = getStack(fullBox!!, emptyChest)
        Assertions.assertTrue(fullChest != null, "Failed to insert boxes in a chest")
        chestOf(fullBox, fullChest!!)
    }

    /**
     * <pre>
     * Main hand: 1 shulker box (1728 diamonds)
     * Off hand: 1 bundle
     * Expected result: 1 bundle (64 * BUNDLE_FILL shulker boxes)</pre>
     */
    @Test
    fun boxInBundle() {
        val testItem = testItem
        val fullBox: ItemStack? = getStack(testItem, ItemStack.EMPTY)
        Assertions.assertTrue(fullBox != null, "Failed to insert item in a box")
        val fullBundle: ItemStack? = getStack(fullBox!!, emptyBundle)
        Assertions.assertTrue(fullBundle != null, "Failed to insert boxes in a chest")
        bundleOf(fullBox, fullBundle!!)
    }

    /**
     * <pre>
     * Main hand: 1 shulker box (1728 diamonds)
     * Off hand: 1 shulker box
     * Expected result: none</pre>
     */
    @Test
    fun boxInBox() {
        val testItem = testItem
        val fullBox: ItemStack? = getStack(testItem, ItemStack.EMPTY)
        Assertions.assertTrue(fullBox != null, "Failed to insert item in a box")
        val fullBox2: ItemStack? = getStack(fullBox!!, getEmptyBox(null))
        Assertions.assertFalse(fullBox2 != null, "Should return no item")
    }

    /**
     * <pre>
     * Main hand: 1 white shulker box
     * Off hand: 1 chest
     * Expected result: 1 chest (1728 shulker boxes)</pre>
     */
    @Test
    fun emptyBoxInChest() {
        val testItem = getEmptyBox(DyeColor.WHITE)
        val fullChest: ItemStack? = getStack(testItem, emptyChest)
        Assertions.assertTrue(fullChest != null, "Failed to insert boxes in a chest")
        chestOf(testItem.copyWithCount(64), fullChest!!)
    }

    /**
     * <pre>
     * Main hand: 1 white shulker box
     * Off hand: none
     * Expected result: 1 chest (1728 shulker boxes)</pre>
     */
    @Test
    fun emptyBox() {
        val testItem = getEmptyBox(DyeColor.WHITE)
        val fullChest: ItemStack? = getStack(testItem, ItemStack.EMPTY)
        Assertions.assertTrue(fullChest != null, "Failed to insert boxes in a chest")
        chestOf(testItem.copyWithCount(64), fullChest!!)
    }

    /**
     * <pre>
     * Main hand: 1 white shulker box
     * Off hand: 1 bundle
     * Expected result: 1 bundle (64 * BUNDLE_FILL shulker boxes)</pre>
     */
    @Test
    fun emptyBoxInBundle() {
        val testItem = getEmptyBox(DyeColor.WHITE)
        val fullBundle: ItemStack? = getStack(testItem, emptyBundle)
        Assertions.assertTrue(fullBundle != null, "Failed to insert boxes in a bundle")
        bundleOf(testItem.copyWithCount(64), fullBundle!!)
    }

    /**
     * <pre>
     * Main hand: 1 white shulker box
     * Off hand: 1 shulker box
     * Expected result: none</pre>
     */
    @Test
    fun emptyBoxInBox() {
        val testItem = getEmptyBox(DyeColor.WHITE)
        val fullBox: ItemStack? = getStack(testItem, getEmptyBox(null))
        Assertions.assertFalse(fullBox != null, "Should return no item")
    }

    /**
     * <pre>
     * Main hand: none
     * Off hand: none
     * Expected result: none</pre>
     */
    @Test
    fun noItem() {
        val fullBox: ItemStack? = getStack(ItemStack.EMPTY, ItemStack.EMPTY)
        Assertions.assertFalse(fullBox != null, "Should return no item")
    }

    /**
     * <pre>
     * Main hand: none
     * Off hand: 1 shulker box
     * Expected result: none</pre>
     */
    @Test
    fun noItemWithOffhand() {
        val fullBox: ItemStack? = getStack(ItemStack.EMPTY, getEmptyBox(null))
        Assertions.assertFalse(fullBox != null, "Should return no item")
    }

    /**
     * <pre>
     * Main hand: 1 chest (1728 diamonds)
     * Off hand: 1 shulker box
     * Expected result: none</pre>
     */
    @Test
    fun chestInBoxInOffhand() {
        val testItem = testItem
        val fullChest: ItemStack? = getStack(testItem, emptyChest)
        Assertions.assertTrue(fullChest != null, "Failed to insert item in a box")
        val fullBox: ItemStack? = getStack(fullChest!!, getEmptyBox(null))
        Assertions.assertFalse(fullBox != null, "Should return no item")
    }

    /**
     * <pre>
     * Main hand: 1 chest (1728 diamonds)
     * Off hand: none
     * Expected result: none</pre>
     */
    @Test
    fun chestInBox() {
        val testItem = testItem
        val fullChest: ItemStack? = getStack(testItem, emptyChest)
        Assertions.assertTrue(fullChest != null, "Failed to insert item in a box")
        val fullBox: ItemStack? = getStack(fullChest!!, ItemStack.EMPTY)
        Assertions.assertFalse(fullBox != null, "Should return no item")
    }

    /**
     * <pre>
     * Main hand: 1 chest (1728 diamonds)
     * Off hand: 1 bundle
     * Expected result: none</pre>
     */
    @Test
    fun chestInBundle() {
        val testItem = testItem
        val fullChest: ItemStack? = getStack(testItem, emptyChest)
        Assertions.assertTrue(fullChest != null, "Failed to insert item in a box")
        val fullBundle: ItemStack? = getStack(fullChest!!, emptyBundle)
        Assertions.assertFalse(fullBundle != null, "Should return no item")
    }

    private fun boxOf(stack: ItemStack, box: ItemStack, color: DyeColor?) {
        Assertions.assertEquals(
            color,
            ShulkerBoxBlock.getColor(box.item),
            "Shulker box color '${ShulkerBoxBlock.getColor(box.item)}' doesn't match expected '$color'"
        )
        Assertions.assertTrue(box.hasNbt(), "Shulker box has no NBT")
        Assertions.assertTrue(
            box.nbt!!.contains("BlockEntityTag", NbtElement.COMPOUND_TYPE.toInt()),
            "Shulker box has no BlockEntityTag"
        )
        val shulker = ShulkerBoxBlockEntity.createFromNbt(
            BlockPos.ORIGIN,
            ShulkerBoxBlock.get(color).defaultState,
            box.getSubNbt("BlockEntityTag")
        ) as ShulkerBoxBlockEntity?
        for (i in 0 until shulker!!.size()) {
            Assertions.assertTrue(
                ItemStack.areEqual(stack, shulker.getStack(i)),
                "Shulker box item '${shulker.getStack(i)}' doesn't match expected '$stack'"
            )
        }
    }

    private fun chestOf(stack: ItemStack, chest: ItemStack) {
        Assertions.assertTrue(chest.hasNbt(), "Chest has no NBT")
        Assertions.assertTrue(
            chest.nbt!!.contains("BlockEntityTag", NbtElement.COMPOUND_TYPE.toInt()),
            "Chest has no BlockEntityTag"
        )
        val container = ChestBlockEntity.createFromNbt(
            BlockPos.ORIGIN,
            Blocks.CHEST.defaultState,
            chest.getSubNbt("BlockEntityTag")
        ) as ChestBlockEntity?
        for (i in 0 until container!!.size()) {
            Assertions.assertTrue(
                ItemStack.areEqual(stack, container.getStack(i)),
                "Chest item '${container.getStack(i)}' doesn't match expected '$stack'"
            )
        }
    }

    private fun bundleOf(stack: ItemStack, bundle: ItemStack) {
        Assertions.assertTrue(bundle.hasNbt(), "Bundle has no NBT")
        val nbtCompound = bundle.getOrCreateNbt()
        Assertions.assertTrue(nbtCompound.contains("Items", NbtElement.LIST_TYPE.toInt()), "Bundle has no Items tag")
        val items = nbtCompound.getList("Items", NbtElement.COMPOUND_TYPE.toInt())
        Assertions.assertFalse(items.isEmpty(), "Bundle is empty")
        for (element in items) {
            val item = ItemStack.fromNbt(element as NbtCompound)
            Assertions.assertTrue(
                ItemStack.areEqual(stack, item),
                "Bundle item '$item' doesn't match expected '$stack'"
            )
        }
    }

    private val testItem: ItemStack
        get() = Items.DIAMOND.defaultStack

    private fun getEmptyBox(color: DyeColor?): ItemStack {
        return ShulkerBoxBlock.getItemStack(color)
    }

    private val emptyChest: ItemStack
        get() = Items.CHEST.defaultStack
    private val emptyDropper: ItemStack
        get() = Items.DROPPER.defaultStack
    private val emptyBundle: ItemStack
        get() = Items.BUNDLE.defaultStack

    companion object {
        @BeforeAll
        @JvmStatic
        fun setup() {
            SharedConstants.createGameVersion()
            Bootstrap.initialize()
        }
    }
}
